src/shared/components/oncoprint/OncoprintUtils.ts

```diff
@@ -301,6 +364,70 @@ export function makeHeatmapTracksMobxPromise(oncoprint:ResultsViewOncoprint, sam
     });
 }
 
+export function makeGenesetHeatmapExpansionsMobxPromise(oncoprint:ResultsViewOncoprint, sampleMode:boolean) {
+    return remoteData<{[genesetTrackKey: string]: IGeneHeatmapTrackSpec[]}>({
+        await: () => [
+            oncoprint.props.store.samples,
+            oncoprint.props.store.patients,
+            oncoprint.props.store.molecularProfileIdToMolecularProfile,
+            oncoprint.props.store.geneMolecularDataCache,
+            oncoprint.props.store.genesetCorrelatedGeneCache
+        ],
+        invoke: async () => {
+            const samples = oncoprint.props.store.samples.result!;
+            const patients = oncoprint.props.store.patients.result!;
+            const molecularProfileIdToMolecularProfile = oncoprint.props.store.molecularProfileIdToMolecularProfile.result!;
+            const dataCache = oncoprint.props.store.geneMolecularDataCache.result!;
+            const genesetGeneCache = oncoprint.props.store.genesetCorrelatedGeneCache.result!;
+
+            const trackGroup = oncoprint.genesetHeatmapTrackGroup;
+            const expansionsByGenesetTrack = oncoprint.genesetHeatmapTrackExpansionGenes;
+
+            // list all the genes in an array of plain, non-observable objects,
+            // as observable arrays cannot be safely passed to external libs
+            const cacheQueries: ({entrezGeneId: number, molecularProfileId: string})[] =
+                _.flatten(expansionsByGenesetTrack.values().map(mobxArray => mobxArray.slice()))
+                .map(({entrezGeneId, molecularProfileId}) => ({entrezGeneId, molecularProfileId}));
+            await dataCache.getPromise(cacheQueries, true);
+
+            const tracksByGenesetTrack: {[genesetTrackKey: string]: IGeneHeatmapTrackSpec[]} = {};
+            expansionsByGenesetTrack.entries().forEach(
+                ([gsTrack, genes]) => {
+                    tracksByGenesetTrack[gsTrack] = genes.map(
+                        ({entrezGeneId, hugoGeneSymbol, molecularProfileId, correlationValue}) => {
+                            const data = dataCache.get({entrezGeneId, molecularProfileId})!.data!;
+                            const profile = molecularProfileIdToMolecularProfile[molecularProfileId];
+                            return {
+                                key: `EXPANSIONTRACK_${gsTrack},${hugoGeneSymbol},GROUP${trackGroup}`,
```

Comment:
the key shouldnt include the trackGroup, otherwise if the track group changes (e.g. "pushed up") then it won't be associated as the same track and will be removed and re-added by `DeltaUtils`

I'd say use the gene set id (is that the molecularProfileId?) instead

Reply:
That's not something I introduce in this PR, it's already in rc. Please log an issue there.

Reply:
See my [other response](https://github.com/cBioPortal/cbioportal-frontend/pull/981#discussion_r171168585) to this review for an explanation on why this logic exists in rc.

Reply:
I see, so you have to force the oncoprint to register this as a new track in order to change its trackGroupIndex.